added helper function that joins vectors of strings

evgeniuz opened the pull request:
### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### Description
Added helper function that joins vectors of strings and tests for it.

Fixes #4640.

evgeniuz committed:
added helper function that joins vectors of strings, refactored suggestions formattingi function to use it

axic commented:
This looks great, thank you!

Can you also use this new helper in some places? One place using it is the "did you mean?" suggestion messages for unknown variable names.

axic commented on libdevcore/CommonData.h:
```diff
@@ -47,6 +49,32 @@ enum class HexPrefix
+	std::string result = boost::algorithm::join(_list, _separator);
+
+	// calling .back() on empty _list is undefined behavior
+	if (!_lastSeparator.empty() && !_list.empty()) {
```
Should this allow empty lists as input?

If yes, should we just have a `if (_list.empty()) return string{}` as the first statement before calling `join`?

chriseth reviewed (changes requested):
Thanks for the pull request!

Could you please update the coding style? We generally use `std::string const& _separator`, braces on a line of their own and tabs for indentation.

Also I think that the case where there is a last separator looks a little hacking. Perhaps it is best to just implement everything manually?

evgeniuz commented:
Updated coding style, refactored function in a less hackish way (it's actually better that way, now it can work with arrays as well, as it doesn't depend on `_list.back()` and `_list.empty()` anymore), also refactored `quotedAlternativesList` to use this new function.

axic commented on libdevcore/CommonData.h:
```diff
@@ -56,20 +54,34 @@ enum class HexPrefix
-			_separator + _list.back(),
-			_lastSeparator + _list.back()
-		);
+std::string join
```
I think this should be called something like `joinHumanReadablelList`. Any better suggestions?

axic replied:
Also likely it should go into the `StringUtils` file.

evgeniuz replied:
ok, will rename and move it right now.

evgeniuz replied:
maybe just joinHumanReadable?

chriseth commented on libdevcore/StringUtils.h:
```diff
@@ -36,4 +36,43 @@ size_t stringDistance(std::string const& _str1, std::string const& _str2);
+std::string joinHumanReadable
+(
+	T const& _list,
+	std::string const& _separator =", ",
```
Please add a space after `=`

chriseth commented on libdevcore/StringUtils.h:
```diff
@@ -36,4 +36,43 @@ size_t stringDistance(std::string const& _str1, std::string const& _str2);
+		++it;
+	}
+
+	for (;it != itEnd; ++it) {
```
Please put the `{` on a line of its own just as with function bodies (also above)

chriseth commented on libdevcore/StringUtils.h:
```diff
@@ -36,4 +36,43 @@ size_t stringDistance(std::string const& _str1, std::string const& _str2);
+		if ((next(it) == itEnd) && !_lastSeparator.empty()) {
+			// last iteration
+			result += _lastSeparator;
+		} else {
```
Please omit braces if there is only a single statement.

chriseth reviewed (changes requested):
Wonderful! There are only some small style changes left.

axic commented:
@evgeniuz as a last step when all the changes are agreed on, can you please squash the commits down to a single one?

You could use `git rebase -i HEAD~4` to do it.

axic commented:
@chriseth do we have any more comments? I am not sure about the function name, but have no good suggestions.
